File: sourcerer-core/src/main/java/org/elder/sourcerer2/DefaultEventSubscription.java
package org.elder.sourcerer2;

import org.elder.sourcerer2.subscription.EventSubscriptionManager;

public class DefaultEventSubscription<T> implements EventSubscription {
    private final EventRepository<T> repository;
    private final Integer shard;
    private final EventSubscriptionHandler<T> subscriptionHandler;
    private final SubscriptionWorkerConfig config;
    private EventSubscriptionPositionSource positionSource;

    public DefaultEventSubscription(
            final EventRepository<T> repository,
            final Integer shard,
            final EventSubscriptionHandler<T> subscriptionHandler,
            final SubscriptionWorkerConfig config
    ) {
        this.repository = repository;
        this.shard = shard;
        this.subscriptionHandler = subscriptionHandler;
        this.config = config;
    }

    @Override
    public void setPositionSource(final EventSubscriptionPositionSource positionSource) {
        this.positionSource = positionSource;
    }

    @Override
    public SubscriptionToken start() {
        EventSubscriptionManager<T> subscriptionManager = new EventSubscriptionManager<>(
                repository,
                shard,
                positionSource,
                subscriptionHandler,
                config);
        return subscriptionManager.start();
    }
}

File: sourcerer-core/src/main/java/org/elder/sourcerer2/DefaultEventSubscription.kt
package org.elder.sourcerer2

import kotlinx.coroutines.CoroutineDispatcher
import org.elder.sourcerer2.subscription.EventSubscriptionManager
import java.lang.IllegalStateException

class DefaultEventSubscription<T>(
        private val dispatcher: CoroutineDispatcher,
        private val repository: EventRepository<T>,
        private val shard: RepositoryShard?,
        private val subscriptionHandler: EventSubscriptionHandler<T>,
        private val config: SubscriptionWorkerConfig
) : EventSubscription {
    private var positionSource: EventSubscriptionPositionSource? = null

    override fun setPositionSource(positionSource: EventSubscriptionPositionSource) {
        this.positionSource = positionSource
    }

    override fun start(): SubscriptionToken {
        val actualSubscriptionPositionSource = positionSource
                ?: throw IllegalStateException(
                        "A subscription position source is required to start a subscription")

        val subscriptionManager = EventSubscriptionManager(
                dispatcher,
                repository,
                shard,
                actualSubscriptionPositionSource,
                subscriptionHandler,
                config)
        return subscriptionManager.start()
    }
}
